File: packages/dirigent-server/src/dirigent_server/routes/workers.py
"""The worker registry: who is alive, on what version, with which plugins and tags."""

import sqlalchemy as sa
from fastapi import APIRouter

from dirigent_client.enums import WorkerStatus
from dirigent_client.schemas import Page, WorkerOut
from dirigent_core.models import Worker, utcnow
from dirigent_core.registry import STALE_AFTER
from dirigent_server.dependencies import ServicesDep, SessionDep
from dirigent_server.pagination import DEFAULT_PAGE, AfterParam, LimitParam, clip
from dirigent_server.security import PrincipalDep
from dirigent_server.transactions import Transactional

router = APIRouter(route_class=Transactional, tags=["workers"])


@router.get("/workers", operation_id="listWorkers", summary="List workers", response_model=Page[WorkerOut])
async def list_workers(
    session: SessionDep,
    services: ServicesDep,
    principal: PrincipalDep,
    after: AfterParam = None,
    limit: LimitParam = DEFAULT_PAGE,
) -> Page[WorkerOut]:
    """List the worker registry, flagging anything stale or running different code."""
    digest = services.host.catalog().digest
    cutoff = utcnow() - STALE_AFTER
    statement = sa.select(Worker).order_by(Worker.name).limit(limit + 1)
    if after is not None:
        statement = statement.where(Worker.name > after)
    rows = await session.execute(statement)
    found = [
        WorkerOut(
            id=row.id,
            name=row.name,
            hostname=row.hostname,
            version=row.version,
            status=row.status,
            concurrency=row.concurrency,
            tags=list(row.tags),
            plugins=dict(row.plugins),
            catalog_digest=row.catalog_digest,
            code_matches_server=row.catalog_digest == digest,
            stale=row.last_seen_at < cutoff and row.status is not WorkerStatus.STOPPED,
            created_at=row.created_at,
            last_seen_at=row.last_seen_at,
        )
        for row in rows.scalars()
    ]
    items, following = clip(found, limit, lambda row: row.name)
    return Page(items=items, next=following)

File: packages/dirigent-server/frontend/src/lib/workers.ts
/**
 * The worker registry, and what one glance at it should say.
 *
 * THE FIELD NAMES ARE THE WIRE'S: `WorkerOut` in `dirigent_client.schemas.system`, member for
 * member. `stale` and `code_matches_server` are the server's own conclusions and are not
 * recomputed here.
 *
 * A TILE STATES THE WORST FACT, NOT THE FIRST ONE. Six workers, one of them running code this
 * server does not have -- the tile that said "5 of 6 answering, worker-a is starting" would be
 * telling the truth and saying nothing. `worstConcern` orders every fact a worker can present
 * and answers the one a person has to know first, and it is a pure function so that ordering is
 * a decision a test makes rather than a judgement a reviewer has to re-make.
 */

import { apiJson, type Page } from '@/lib/api'
import { PAGE } from '@/lib/paging'

/** Where a worker is in its own life. `WorkerStatus` in dirigent_client.enums. */
export type WorkerStatus = 'starting' | 'running' | 'draining' | 'stopped'

/** One node claiming work. `WorkerOut`. */
export interface WorkerOut {
    id: string
    name: string
    hostname: string
    version: string
    status: WorkerStatus
    concurrency: number
    tags: string[]
    plugins: Record<string, unknown>
    catalog_digest: string | null
    /** Whether this worker's block catalog is the one the server has. */
    code_matches_server: boolean
    /** Whether the server has stopped hearing from it. Computed server-side against its own clock. */
    stale: boolean
    created_at: string
    last_seen_at: string
}

/** Where one page of the workers listing is read from. */
export function workersPath(after: string | null, limit: number = PAGE): string {
    const query = new URLSearchParams({ limit: String(limit) })
    if (after !== null) query.set('after', after)
    return `/workers?${query.toString()}`
}

/** Read one page of workers, in name order. */
export function readWorkers(after: string | null): Promise<Page<WorkerOut>> {
    return apiJson<Page<WorkerOut>>(workersPath(after))
}

/**
 * Whether any worker in the registry carries a tag, which is what makes the Tags column worth
 * a column. An instance that never routes by tag would read a column of dashes otherwise.
 */
export function anyTagged(workers: readonly WorkerOut[]): boolean {
    return workers.some((worker) => worker.tags.length > 0)
}

/** Whether a worker is answering: heard from recently, and not shut down. */
export function workerAlive(worker: WorkerOut): boolean {
    return !worker.stale && worker.status !== 'stopped'
}

/** What one worker presents, worst first. `null` is a worker with nothing to report. */
export type Concern = 'silent' | 'mismatched' | 'draining' | 'stopped' | 'starting'

/**
 * The concerns in the order a person needs to hear them.
 *
 * A worker nobody has heard from may be holding a claimed attempt that will never finish, so it
 * is first. One that answers with the wrong catalog is second: it is up, and it will run
 * something other than what this server thinks it will. Draining is capacity leaving now,
 * stopped is capacity already gone and said so, starting is nothing yet.
 */
export const CONCERNS: readonly Concern[] = ['silent', 'mismatched', 'draining', 'stopped', 'starting']

/** What one worker presents, or nothing. */
export function concernOf(worker: WorkerOut): Concern | null {
    if (worker.stale) return 'silent'
    if (!worker.code_matches_server) return 'mismatched'
    if (worker.status === 'draining') return 'draining'
    if (worker.status === 'stopped') return 'stopped'
    if (worker.status === 'starting') return 'starting'
    return null
}

/** One worker's concern as a sentence naming it. */
export function concernNote(worker: WorkerOut, concern: Concern): string {
    return `${worker.name} ${concernSaid(concern)}`
}

/** What is wrong, for a row already headed by the worker it is wrong with. */
export function concernSaid(concern: Concern): string {
    switch (concern) {
        case 'silent':
            return 'has gone quiet'
        case 'mismatched':
            return 'is running a different catalog from this server'
        case 'draining':
            return 'is draining'
        case 'stopped':
            return 'has stopped'
        case 'starting':
            return 'is still starting'
    }
}

/** The one worker whose fact is worst, and which fact it is. Nothing when every worker is well. */
export function worstConcern(workers: readonly WorkerOut[]): { worker: WorkerOut; concern: Concern } | null {
    let worst: { worker: WorkerOut; concern: Concern; rank: number } | null = null
    for (const worker of workers) {
        const concern = concernOf(worker)
        if (concern === null) continue
        const rank = CONCERNS.indexOf(concern)
        if (worst === null || rank < worst.rank) worst = { worker, concern, rank }
    }
    return worst === null ? null : { worker: worst.worker, concern: worst.concern }
}
